Empire/Empire.cpp

```diff
@@ -855,6 +920,9 @@ float ProductionQueue::TotalPPsSpent() const {
     return retval;
 }
 
+/** TODO: Is there any reason to keep this method in addition to the more specific 
+ * information directly available from the empire?  This should probably at least be renamed
+ * to clarify it is non-stockpile output */
```

Comment:
Keep in mind that this method is returning a ResourceGroup:amount mapping of available PP, not a simple tally, which is provided via ResourcePool::TotalAvailable().  And there we clearly identify TotalAvailable as including stockpile amounts, versus TotalOutput which does not.  This current method is really a repackaging of ResourcePool::Available() plus a check for invalid ResourcePool.  But also looking at the ResourcePool level I see that this does in fact include the stockpile amount, though in a location-specific fashion.  I'll need to look into this a bit more, I suspect I've been misusing this a bit because of that stockpile issue.

Which is making me feel like that is altogether more reason to eliminate this method from here and just rely directly on the underlying ResourcePool::Available() -- extra layers of low-value indirection appear to be ripe for me to make errors with.

In either place, though, there is an issue of how to report the stockpile info.  It looks to me like the preexisting code was set up to include the stockpile amount in the resource group that contains the stockpile location.  If we move away from a stockpile location then the stockpile amount, whether provided via this or (if we dropped this method) via the underlying ResourcePool::Available() method, would have to be associated with some special set<int>.  I have a recollection that the empty set is used for some placeholder purposes in some of the related code and so I think it best not to use that for this.  So that leaves the most simple being a set including only Invalid_Object_ID (which is kind of congruent with some of the preexisting stockpile code that basically takes INVALID_OBJECT_ID as the default stockpile location); another possibility instead of Invalid_Object_ID is to designate a new special object ID to be used for stockpile where a location ID is needed.  agrrr3 had suggested the value -4 for that purpose, and I can see some appeal to using a new special ID value like -4.